From 942fdcf00e8c160eb6f0f1d28b5b92289b666f4d Mon Sep 17 00:00:00 2001
From: Hui Min Mina Chou <minachou@andestech.com>
Date: Mon, 20 May 2024 11:21:46 +0800
Subject: [PATCH 051/168] ftsdc010g: Block CMD48 which ftsdc010 controller not
 support

ftsdc010 controller only supports SD spec v2. From kernel v5.14 SD spec v4
is supported[1], and CMD48 is sent during SD card initialization[2], causing
SD card recognition failure. So, we block sending of CMD48.

fio result on AX65_4C with a T-CREATE 256G S.M.A.R.T. UHS-I U3 A2 V30 MicroSDXC Card[3]:
  Run status group 0 (all jobs):
     READ: bw=2191KiB/s (2244kB/s), 2191KiB/s-2191KiB/s(2244kB/s-2244kB/s), io=388MiB (407MB), run=181158-181158msec
    WRITE: bw=2259KiB/s (2313kB/s), 2259KiB/s-2259KiB/s(2313kB/s-2313kB/s), io=400MiB (419MB), run=181158-181158msec

[1]: https://github.com/torvalds/linux/commit/dbea8ae9febdea11cb74d094e6b730987079679e
[2]: https://github.com/torvalds/linux/blob/eb6a9339efeb6f3d2b5c86fdf2382cdc293eca2c/drivers/mmc/core/sd.c#L1260
[3]: https://www.teamgroupinc.com/en/product-detail/card/T-CREATE/expert-smart/expert-smart-TTCS256GIA2V3003/
Signed-off-by: Hui Min Mina Chou <minachou@andestech.com>
---
 drivers/mmc/host/ftsdc010g.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/mmc/host/ftsdc010g.c b/drivers/mmc/host/ftsdc010g.c
index 19b39c9318c1..26d933afb012 100644
--- a/drivers/mmc/host/ftsdc010g.c
+++ b/drivers/mmc/host/ftsdc010g.c
@@ -11,6 +11,7 @@
 #include <linux/clk.h>
 #include <linux/mmc/host.h>
 #include <linux/mmc/mmc.h>
+#include <linux/mmc/sd.h>
 #include <linux/mmc/card.h>
 #include <linux/platform_device.h>
 #include <linux/debugfs.h>
@@ -965,6 +966,17 @@ static void ftsdc_send_request(struct mmc_host *mmc)
 	prepare_dbgmsg(host, cmd, host->cmd_is_stop);
 	dbg(host, dbg_debug, "%s\n", host->dbgmsg_cmd);
 
+	/*
+	 * ftsdc010 controller only supports SD spec v2. From kernel v5.14 SD spec v4
+	 * is supported, and CMD48 is sent during SD card initialization, causing
+	 * SD card recognition failure. So, we block CMD48 here.
+	 */
+	if (cmd->opcode == SD_READ_EXTR_SINGLE) {
+		dbg(host, dbg_info, "Skip sending CMD%d\n", SD_READ_EXTR_SINGLE);
+		mmc_request_done(mmc, mrq);
+		return;
+	}
+
 	if (cmd->data) {
 		int res = ftsdc_setup_data(host, cmd->data);
 
-- 
2.34.1

